### DISCUSSION TITLE
How to set the value of kappa as a function of X and Y

### DISCUSSION THREAD

**Discussion by biaogxb** [idaholab/moose#22013] [Q&A Modules: Phase field] [answered]
Hello, I want to define the value of kappa as a function, as shown below

```
[Kernels]
  [./AC_int]
    type = ACInterface
    variable = eta
    kappa_name = kappa_op
  [../]
[Materials]
 [./kappa_op]
    type = ParsedMaterial
    f_name = kappa_op 
    args = 'x y' 
    function = '2.4+0.12*cos(4*atan2(y,x))'
  [../]
```

However, the following error occurred:
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'kappa_op', requested by 'AC_int' is not defined on block 0

I would like to ask how to correctly set kappa as a function about X and Y. thank you！！！

**Comment by GiudGiud** [comment, marked as answer]
Hello

I dont think you need to specify x and y in args, you can just leave them in function
The name should have been properly set though. Can you please attach your input file?

Guillaume

**Reply by biaogxb** [reply]
Thank you for your reply! I want to implement kappa as a function about X and Y. I have added a series of # annotations to the involved parts in the input file so that you can find it more quickly.


I'm not sure if I can set kappa as a function and operate successfully through my current operation. I tried it. At present, the convergence is not very good. Do you agree with the way I modify this code or do you have a simpler and more correct way to realize the functionalization of kappa?

I want to set kappa = 2.4 (1 + 0.05 * cos (4* θ））
According to the following formula:
![image](https://user-images.githubusercontent.com/95063954/188775340-7e8c0948-3a76-4366-9042-5d62c476108a.png)
![image](https://user-images.githubusercontent.com/95063954/188775387-7f65a688-ecee-41bc-8d04-24435935b541.png)
![image](https://user-images.githubusercontent.com/95063954/188775267-ef3f7ea7-5fde-4415-980a-c24925479232.png)

this θ It is an angle that can be represented by 'x' and 'y', which will change at any time. Therefore, setting kappa as a constant does not satisfy the condition.
I'm looking forward to your reply. Thank you！：）

**Reply by GiudGiud** [reply]
why do you say x and y will change during time? is the mesh going to displace?

you could probably avoid the auxvariable here, and define it directly in the parsed material.

**Reply by biaogxb** [reply]
θ is the angle between the surface normal vector and the crystallographic orientation.So I modified the code according to my own idea. I want to express this through the coordinates (x, y) of the interface position θ Angle that changes as the surface grows. but obviously I was wrong. How can I express this θ ?

![image](https://user-images.githubusercontent.com/95063954/188814814-49d2642e-c9fc-4551-a6c1-ec7562da6eb4.png)
![image](https://user-images.githubusercontent.com/95063954/188814884-f11fd1f2-28f2-465a-8a36-6da0c9ecbe12.png)
![image](https://user-images.githubusercontent.com/95063954/188814913-efd72c2f-eb4c-4ef1-a495-a3a8ce573610.png)

**Reply by GiudGiud** [reply]
So are you modeling the surface explicitly? Is theta a variable then? Or does it never change at a given x,y during the simulation

**Reply by GiudGiud** [reply]
@laagesen for helping on phase field

**Reply by laagesen** [reply]
The angle of the surface normal vector is not the same as the way you have drawn theta here. To see this, just draw a normal vector to the surface at the end of one of your arrows and you will see it doesn't point in the same direction as the arrow. (If it did, the outer surface that the arrows point to would be a circle instead of the more complex wavy shape).

Because of this, you can't calculate kappa as a function of x and y in your domain. The process is unfortunately more tricky, you will need to calculate the interfacial normal as a function of the gradients of the order parameter. To see an example of how this is done, you can look at  the code in `moose/modules/phase_field/src/materials/InterfaceOrientationMaterial.C`. You can see how this material is used in a dendritic growth example by running `moose/modules/phase_field/examples/anisotropic_interfaces/snow.i`.

**Reply by biaogxb** [reply]
> 

Thank you for your answer. I will study this example!

**Reply by biaogxb** [reply]
> The angle of the surface normal vector is not the same as the way you have drawn theta here. To see this, just draw a normal vector to the surface at the end of one of your arrows and you will see it doesn't point in the same direction as the arrow. (If it did, the outer surface that the arrows point to would be a circle instead of the more complex wavy shape).
> 
> Because of this, you can't calculate kappa as a function of x and y in your domain. The process is unfortunately more tricky, you will need to calculate the interfacial normal as a function of the gradients of the order parameter. To see an example of how this is done, you can look at the code in `moose/modules/phase_field/src/materials/InterfaceOrientationMaterial.C`. You can see how this material is used in a dendritic growth example by running `moose/modules/phase_field/examples/anisotropic_interfaces/snow.i`.

I carefully read the snowflake example and found that the ‘InterfaceOrientationMaterial’ is used to calculate the kappa（eps in that article） value I need, so I changed the code for calculating the interface as follows

`[./anisoACinterface]
    type = ACInterface
    variable = eta
    mob_name = M0
    kappa_name=eps
  [../]`
`[./material]
    type = InterfaceOrientationMaterial
    op = eta
  [../]`
I don't know why my calculation is not convergent. Can't I directly fill ‘eps’ into ACInterface as ‘kappa’？
Here is my input file：
[input.txt](https://github.com/idaholab/moose/files/9971295/input.txt)

**Reply by laagesen** [reply]
I see you are using some logarithmic terms in your free energy expressions. Those are often problematic in terms of convergence. I would suggest first removing the anisotropy and make sure your model works without it first. If you still have problems without the anisotropy I would be suspicious of the logarithmic terms. You can try using the plog function instead, which substitutes a polynomial function below a user-specified tolerance, or try a parabolic approximation to the logarithmic expression.

**Comment by dschwen** [comment]
`ParsedMaterial` cannot use `x` and `y`. Use `GenericFunctionMaterial` instead.

**Reply by biaogxb** [reply]
Thank you for your answer！
